pkg/compact/compact.go

```diff
@@ -309,18 +342,81 @@ func (g *DefaultGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (res []*Gro
 			return nil, errors.Wrap(err, "add compaction group")
 		}
 	}
+
+	if g.planner != nil {
+		newGroups := []*Group{}
+		// split each group by available plans
+		for _, group := range res {
+			newGroup, err := g.splitGroupByPlans(group)
+			if err != nil {
+				return nil, err
+			}
+			newGroups = append(newGroups, newGroup...)
+		}
+		res = newGroups
+	}
+
 	sort.Slice(res, func(i, j int) bool {
 		return res[i].Key() < res[j].Key()
 	})
 	return res, nil
 }
 
+// Recursively plan a groups metas until there are no available plans and return a new group for each plan
+// This allows a single group key to utilize group concurrency when there are multiple plans available to compact
+func (g *DefaultGrouper) splitGroupByPlans(group *Group) ([]*Group, error) {
+	groups := []*Group{}
+	// split each group by available plans
+	var seededRand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	for toCompact, err := g.planner.Plan(context.Background(), group.metasByMinTime); len(toCompact) != 0; toCompact, err = g.planner.Plan(context.Background(), group.metasByMinTime) {
+		if err != nil {
+			return nil, errors.Wrap(err, "split group")
+		}
+		lbls := group.Labels()
+		hash := strconv.Itoa(seededRand.Int())
+		groupKey := group.Key()
+		newGroup, err := NewGroup(
```

Comment:
Can we reuse group? This feel a bit wrong that we create group only to then split by group and create again 🤗

Reply:
Yeah this can be refactored to not use group, all thats really needed is the group key and the metasByMinTime

Reply:
After second look reusing group seems like the most efficient way. If you would rather i can make a new type like GroupSplitter or something like that but it will have some redundancy because i will need to duplicate the metasByMinTime sorting and some other fields. Happy to change it if thats preferred